### PR TITLE
Rework Observers to use collision free map

### PR DESCRIPTION
**Pull request by keith-turner**

### CODE REVIEW

File: modules/data/src/test/java/io/fluo/webindex/data/fluo/IndexIT.java [hunk 1]
```diff
@@ -163,7 +167,8 @@ private void assertOutput(Collection<Page> pages, FluoClient client) throws Exce
     // Create expected output using spark
     IndexStats stats = new IndexStats(sc);
     JavaPairRDD<RowColumn, Bytes> accumuloIndex = IndexUtil.createAccumuloIndex(stats, pagesRDD);
-    JavaPairRDD<RowColumn, Bytes> fluoIndex = IndexUtil.createFluoIndex(accumuloIndex);
+    JavaPairRDD<RowColumn, Bytes> fluoIndex =
+        IndexUtil.createFluoIndex(accumuloIndex, 5).sortByKey();
```
Comment (mikewalch) [inline modules/data/src/test/java/io/fluo/webindex/data/fluo/IndexIT.java]: How about changing the 5 to `FluoApp.NUM_BUCKETS`?

Comment (mikewalch) [inline modules/data/src/test/java/io/fluo/webindex/data/fluo/IndexIT.java]: The `sortByKey` should be removed.

File: modules/data/src/test/java/io/fluo/webindex/data/fluo/IndexIT.java [hunk 2]
```diff
@@ -124,7 +126,9 @@ public void setUpFluo() throws Exception {
     exportTable = "export" + tableCounter.getAndIncrement();
     cluster.getConnector("root", "secret").tableOperations().create(exportTable);
 
-    PrintProps.configureApplication(config,
+    config.addObserver(new ObserverConfiguration(PageObserver.class.getName()));
+
+    FluoApp.configureApplication(config,
```
Comment (mikewalch) [inline modules/data/src/test/java/io/fluo/webindex/data/fluo/IndexIT.java]: 5 could could set to `FluoApp.NUM_BUCKETS`

File: modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java [hunk 3]
```diff
@@ -160,10 +169,86 @@ private static void addRCV(List<Tuple2<RowColumn, Long>> tuples, String r, Colum
    * Creates a Fluo index by filtering out unnecessary data from Accumulo Index
    */
   public static JavaPairRDD<RowColumn, Bytes> createFluoIndex(
-      JavaPairRDD<RowColumn, Bytes> accumuloIndex) {
+      JavaPairRDD<RowColumn, Bytes> accumuloIndex, int numBuckets) {
+
+    KryoSimplerSerializer serializer = new KryoSimplerSerializer(new WebindexKryoFactory());
+
     JavaPairRDD<RowColumn, Bytes> fluoIndex =
```
Comment (mikewalch) [inline modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java]: This code looks like it should be removed

File: modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java [hunk 4]
```diff
@@ -160,10 +169,86 @@ private static void addRCV(List<Tuple2<RowColumn, Long>> tuples, String r, Colum
...
+
+          return new Tuple2<RowColumn, Bytes>(new RowColumn(rcv.getRow(), rcv.getColumn()), rcv
+              .getValue());
+        });
+
+    Initializer<String, Long> domainMapInitializer =
+        CollisionFreeMap.getInitializer(DomainMap.DOMAIN_MAP_ID, numBuckets, serializer);
+
+    // generate the rest of the fluo table
+    fluoIndex =
```
Comment (mikewalch) [inline modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java]: If the above code is removed, the variable type needs to be declared here.

File: modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java [hunk 5]
```diff
@@ -160,10 +169,86 @@ private static void addRCV(List<Tuple2<RowColumn, Long>> tuples, String r, Colum
...
+                return Collections.singleton(new Tuple2<RowColumn, Bytes>(new RowColumn(rcv
+                    .getRow(), rcv.getColumn()), rcv.getValue()));
+              }
+              if (row.startsWith("p:") && cf.equals(Constants.PAGE) && cq.equals(Constants.CUR)) {
+                return Collections.singleton(t);
+              }
+
+              return Collections.emptyList();
+            }).union(uriMap);
+
```
Comment (mikewalch) [inline modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java]: `fluoIndex` could be sorted here

File: modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java (hunk 3 above)
Comment (keith-turner) [inline modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java]: Nice catch.

File: modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java (hunk 5 above)
Comment (keith-turner) [inline modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java]: MAde a change in 8378d84.  Kinda shooting in the dark w/ the persist stuff

File: modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java (hunk 4 above)
Comment (keith-turner) [inline modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java]: fixed in 8378d84

File: modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java (hunk 3 above)
Comment (keith-turner) [inline modules/data/src/main/java/io/fluo/webindex/data/spark/IndexUtil.java]: fixed in 8378d84

File: modules/data/src/test/java/io/fluo/webindex/data/fluo/IndexIT.java (hunk 2 above)
Comment (keith-turner) [inline modules/data/src/test/java/io/fluo/webindex/data/fluo/IndexIT.java]: fixed in 8378d84

File: modules/data/src/test/java/io/fluo/webindex/data/fluo/IndexIT.java (hunk 1 above)
Comment (keith-turner) [inline modules/data/src/test/java/io/fluo/webindex/data/fluo/IndexIT.java]: fixed in 8378d84

File: modules/data/src/main/java/io/fluo/webindex/data/fluo/DomainMap.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+import io.fluo.recipes.export.ExportQueue;
+import io.fluo.recipes.map.CollisionFreeMap;
+import io.fluo.recipes.map.CollisionFreeMap.Options;
+import io.fluo.recipes.map.Combiner;
+import io.fluo.recipes.map.Update;
+import io.fluo.recipes.map.UpdateObserver;
+import io.fluo.webindex.data.FluoApp;
+import io.fluo.webindex.data.recipes.Transmutable;
+
+public class DomainMap {
```
Comment (mikewalch) [inline modules/data/src/main/java/io/fluo/webindex/data/fluo/DomainMap.java]: It's cool how you put both the `Combiner` and `UpdateObserver` implementations in the same class.  It makes the code easier to read. You could update the CFM documentation to follow this pattern.

**Comment by mikewalch** [comment]
+1

**Pull request merged**